cmd/snap-confine/snap-confine.c

```diff
@@ -96,6 +97,157 @@ static void sc_maybe_fixup_udev(void)
+ * working directory is that not all paths that exist on the host filesystem
+ * exist in the execution environment. When the current working directory
+ * cannot be restored then a special substitute of /var/lib/snapd/void is used
+ * instead.
+**/
+typedef struct sc_preservation_context {
+	mode_t orig_umask;
+	int orig_cwd_fd;
+} sc_preservation_context;
+
+/**
+ * sc_remember_preservation context stores values to restore later.
+ *
+ * In addition the function makes the following changes to the process:
+ *  - the umask is set to 0
+ *  - the current working directory is set to /
+**/
+static void sc_remember_preservation_context(sc_preservation_context * ctx)
+{
+	/* Reset umask to zero, storing the old value. */
+	ctx->orig_umask = umask(0);
+	debug("umask reset, old umask was %#4o", ctx->orig_umask);
+	/* Remember a file descriptor corresponding to the original working
+	 * directory. This is an O_PATH file descriptor. The descriptor is
+	 * used as explained below. */
+	ctx->orig_cwd_fd =
+	    openat(AT_FDCWD, ".", O_PATH | O_CLOEXEC | O_NOFOLLOW);
+	if (ctx->orig_cwd_fd < 0) {
+		die("cannot open path of the current working directory");
+	}
+	if (chdir("/") < 0) {
+		die("cannot move to /");
+	}
+}
+
+/**
+ *  sc_restore_preservation_context restores values stored earlier.
+**/
+static void sc_restore_preservation_context(const sc_preservation_context * ctx)
+{
+	/* Restore original umask */
+	umask(ctx->orig_umask);
+	debug("umask restored to %#4o", ctx->orig_umask);
+
+	/* Restore original current working directory.
+	 *
+	 * This part is more involved for the following reasons. While we hold an
+	 * O_PATH file descriptor that still points to the original working
+	 * directory, that directory may not be representable in the target mount
+	 * namespace. A quick example may be /custom that exists on the host but
+	 * not in the base snap of the application.
+	 *
+	 * If the path of the original working directory now maps to a different
+	 * inode we cannot use fchdir(2). One example of that is the /tmp
+	 * directory, which exists in both the host mount namespace and the
+	 * per-snap mount namespace but actually represents a different directory.
+	 *
+	 * if that directory is represented it may no longer point to the same
+	 * inode. For example the /tmp directory as it exists on the host is not
+	 * represented inside the execution environment. The /tmp directory inside
+	 * the snap is a specific sub-directory of the host's /tmp directory.
+	 **/
+
+	/* Read the target of symlink at /proc/self/<fd-of-orig-cwd> */
+	char fd_path[PATH_MAX];
+	char orig_cwd[PATH_MAX];
+	ssize_t nread;
+	sc_must_snprintf(fd_path, sizeof fd_path, "/proc/self/fd/%d",
+			 ctx->orig_cwd_fd);
+	nread = readlink(fd_path, orig_cwd, sizeof orig_cwd);
+	if (nread < 0) {
+		die("cannot read symbolic link target %s", fd_path);
+	}
+	if (nread == sizeof orig_cwd) {
+		die("cannot fit symbolic link target %s", fd_path);
+	}
+
```

Comment:
For the sake of debugging, I'd add:
```
debug("original cwd in parent namespace: %s", orig_cwd);
```

Reply:
We cannot do that, that string is not terminated.

Reply:
I meant adding it below:
```
	if (nread == sizeof orig_cwd) {
		die("cannot fit symbolic link target %s", fd_path);
	}
	debug("original cwd in parent namespace: %s", orig_cwd);
```

Reply:
Ah, but that is logged in several cases below?